src/librustc_lint/builtin.rs

```diff
@@ -2196,7 +2198,56 @@ impl ClashingExternDeclarations {
                 | (Opaque(..), Opaque(..)) => false,
                 // These definitely should have been caught above.
                 (Bool, Bool) | (Char, Char) | (Never, Never) | (Str, Str) => unreachable!(),
-                _ => false,
+
+                // Disjoint kinds.
+                (_, _) => {
+                    // First, check if the conversion is FFI-safe. This can be so if the type is an
+                    // enum with a non-null field (see improper_ctypes).
+                    let is_primitive_or_pointer =
+                        |ty: Ty<'tcx>| ty.is_primitive() || matches!(ty.kind, RawPtr(..));
+                    if (is_primitive_or_pointer(a) || is_primitive_or_pointer(b))
+                        && !(is_primitive_or_pointer(a) && is_primitive_or_pointer(b))
+                        && (matches!(a_kind, Adt(..)) || matches!(b_kind, Adt(..)))
+                    /* ie, 1 adt and 1 primitive */
+                    {
+                        let (primitive_ty, adt_ty) =
+                            if is_primitive_or_pointer(a) { (a, b) } else { (b, a) };
+                        // First, check that the Adt is FFI-safe to use.
+                        use crate::types::{ImproperCTypesMode, ImproperCTypesVisitor};
+                        let vis =
+                            ImproperCTypesVisitor { cx, mode: ImproperCTypesMode::Declarations };
+
+                        if let Adt(def, substs) = adt_ty.kind {
+                            let repr_nullable = vis.is_repr_nullable_ptr(adt_ty, def, substs);
+                            if let Some(safe_ty) = repr_nullable {
+                                let safe_ty_layout = &cx.layout_of(safe_ty).unwrap();
+                                let primitive_ty_layout = &cx.layout_of(primitive_ty).unwrap();
+
+                                use rustc_target::abi::Abi::*;
+                                match (&safe_ty_layout.abi, &primitive_ty_layout.abi) {
+                                    (Scalar(safe), Scalar(primitive)) => {
+                                        // The two types are safe to convert between if `safe` is
+                                        // the non-zero version of `primitive`.
+                                        use std::ops::RangeInclusive;
+
+                                        let safe_range: &RangeInclusive<_> = &safe.valid_range;
+                                        let primitive_range: &RangeInclusive<_> =
+                                            &primitive.valid_range;
+
+                                        return primitive_range.end() == safe_range.end()
+                                            // This check works for both signed and unsigned types due to wraparound.
+                                            && *safe_range.start() == 1
+                                            && *primitive_range.start() == 0;
```

Comment:
These two conditions look somewhat sketchy. In the future exactly the same equivalence optimization could apply for `usize == Option<NonAllBitsSetUsize>` where `Option::None` would be represented by `!0`. And this little piece of code would most likely end up being missed during implementation of such functionality. I wonder if there isn’t a better way to check for this…

If not, maybe just compare the size of the two types?

Reply:
> I wonder if there isn’t a better way to check for this…

I've opted for a more general check of seeing if one range excludes a single value of the other, which should also cover the case you've described.

Reply:
Hm… I think I did a bad job at conveying the point I was trying to make here.

What I really wanted to say is that this introduces another source of truth on how the optimisation behaves and where it applies. There is already some code that codifies this kind of information in the layout calculation code (although not easily reusable here, I think) and  `is_repr_nullable_ptr` is yet another place where this happens. _If_, somebody was adjusting how this optimisation works to e.g. admit more cases or fix a bug… there is no provision to ensure that they do not miss updating all of those portions of code, including here. So its better to have fewer such sources of truth than more.

In particular to me it looks like `is_repr_nullable_ptr` already goes to some length to make sure that the contained field is a type that participates in nullable type optimisation, so I suspect this check could be dropped entirely. Is there perhaps a specific reason you want to check ranges that I’m missing? Maybe it would then make sense to move that check to more shared code like `is_repr_nullable_ptr` in the first place, then?

Reply:
>its better to have fewer such sources of truth than more

Ohh, I see what you mean. Yeah, I agree -- that makes sense.

> Is there perhaps a specific reason you want to check ranges that I’m missing?

`is_repr_nullable_ptr` only checks if something like Option<NonZeroUsize> is okay, but doesn't check the type that it's casted to/from. I check the ranges to ensure, for example, that `Option<NonZeroUsize>` is being correctly converted to `usize`, and not to `isize` or another same-sized-but-incorrect type.

> move that check to more shared code like is_repr_nullable_ptr

I've currently got `is_repr_nullable_ptr` just returning the type inside the Option-like enum, which I've come to realise isn't actually that useful. Yeah, I'll go with this -- it'd probably be better to have `is_repr_nullable_ptr` return the type that the `Option<T>` is FFI-convertible to (if it exists), and ClashingExternDeclarations can just compare `primitive_ty` to that type.